src/walk.rs

```diff
@@ -329,224 +212,374 @@ impl<W: Write> ReceiverBuffer<W> {
     }
 }
 
-fn spawn_receiver(
-    config: &Arc<Config>,
-    quit_flag: &Arc<AtomicBool>,
-    interrupt_flag: &Arc<AtomicBool>,
-    rx: Receiver<WorkerResult>,
-) -> thread::JoinHandle<ExitCode> {
-    let config = Arc::clone(config);
-    let quit_flag = Arc::clone(quit_flag);
-    let interrupt_flag = Arc::clone(interrupt_flag);
-
-    let threads = config.threads;
-    thread::spawn(move || {
-        // This will be set to `Some` if the `--exec` argument was supplied.
-        if let Some(ref cmd) = config.command {
-            if cmd.in_batch_mode() {
-                exec::batch(rx, cmd, &config)
-            } else {
-                let out_perm = Mutex::new(());
+/// State shared by the sender and receiver threads.
+struct WorkerState {
+    /// The paths to scan.
+    paths: Vec<PathBuf>,
+    /// The search patterns.
+    patterns: Arc<Vec<Regex>>,
```

Comment:
I wonder if it would be worth combining these Arcs into a single Arc, and pass that around, so we don't need to clone each individually.

Reply:
Even better: https://github.com/sharkdp/fd/pull/1413/commits/ed6396d9be4e41f07167899863045d083c3c5201 just removes the Arcs